File: oldsrc/src/kingdom/dialogs/army-tactics-browser.ts
import {CheckDialog} from './check-dialog';
import {DegreeOfSuccess} from '../../degree-of-success';
import {getArmyTactics, isArmyTactic} from '../../armies/utils';
import {createUUIDLink, getLevelBasedDC, hasArmyTactic} from '../../utils';
import {Kingdom} from '../data/kingdom';

interface ArmyTacticsBrowserOptions {
    game: Game;
    army: Actor & ArmyActor;
    kingdom: Kingdom;
    sheetActor: Actor;
    onRoll: (consumeModifiers: Set<string>) => Promise<void>;
}

interface ArmyTacticView {
    link: string;
    name: string;
    level: number;
    dc: number;
    enabled: boolean;
    uuid: string;
}

interface ArmyBrowserData {
    tactics: ArmyTacticView[];
}

class ArmyTacticsBrowserApp extends FormApplication<
    FormApplicationOptions & ArmyTacticsBrowserOptions,
    object,
    null
> {
    private game: Game;
    private army: Actor & ArmyActor;
    private kingdom: Kingdom;
    private sheetActor: Actor;
    private onRoll: (consumeModifiers: Set<string>) => Promise<void>;

    static override get defaultOptions(): FormApplicationOptions {
        const options = super.defaultOptions;
        options.id = 'army-tactics-browser-app';
        options.title = 'Army Tactics Browser';
        options.template = 'modules/pf2e-kingmaker-tools/templates/kingdom/army-tactics-browser.hbs';
        options.classes = ['kingmaker-tools-app', 'army-tactics-browser-app'];
        options.height = 'auto';
        options.submitOnChange = true;
        options.closeOnSubmit = false;
        return options;
    }

    constructor(options: Partial<ApplicationOptions> & ArmyTacticsBrowserOptions) {
        super(null, options);
        this.game = options.game;
        this.army = options.army;
        this.kingdom = options.kingdom;
        this.sheetActor = options.sheetActor;
        this.onRoll = options.onRoll;
    }

    override async getData(): Promise<ArmyBrowserData> {
        const tactics = await getArmyTactics(this.game);
        return {
            tactics: await this.toViewTactics(tactics),
        };
    }

    override activateListeners(html: JQuery): void {
        super.activateListeners(html);
        const $html = html[0];
        $html.querySelectorAll('.km-tactic-link')
            .forEach(el => el.addEventListener('click', async (ev): Promise<void> => {
                ev.preventDefault();
                ev.stopPropagation();
                const target = ev.currentTarget as HTMLElement;
                const uuid = target.dataset.uuid!;
                const item = await fromUuid(uuid) as Item | null;
                item?.sheet?.render(true);
            }));
        $html.querySelectorAll('.km-train-tactic')
            .forEach(el => el.addEventListener('click', (ev) => this.trainTactic(ev)));
    }

    /* eslint-disable @typescript-eslint/no-explicit-any */
    protected async _updateObject(event: Event, formData: any): Promise<void> {
        console.log(formData);
        this.render();
    }

    private async trainTactic(ev: Event): Promise<void> {
        const button = ev.currentTarget as HTMLElement;
        const id = button.dataset.id!;
        const item: Item<unknown> | null = await fromUuid(id) as Item<unknown> | null;
        if (item !== null && isArmyTactic(item)) {
            const tacticLink = await TextEditor.enrichHTML(createUUIDLink(item.uuid));
            new CheckDialog(null, {
                activity: 'train-army',
                kingdom: this.kingdom,
                dc: getLevelBasedDC(item.system.level.value),
                game: this.game,
                type: 'activity',
                onRoll: this.onRoll,
                actor: this.sheetActor,
                afterRoll: async (): Promise<void> => {
                    await this.close();
                },
                additionalChatMessages: [{
                    [DegreeOfSuccess.CRITICAL_SUCCESS]: tacticLink,
                    [DegreeOfSuccess.SUCCESS]: tacticLink,
                }],
            }).render(true);
        }
    }

    private async toViewTactics(tactics: CampaignFeaturePF2E[]): Promise<ArmyTacticView[]> {
        return await Promise.all(tactics
            .filter(t => t.system.traits.value.includes(this.army.system.traits.type) && !hasArmyTactic(this.army, t))
            .sort((a, b) => a.system.level.value - b.system.level.value || a.name.localeCompare(b.name))
            .map(async (t) => {
                const level = t.system.level.value;
                return {
                    enabled: level <= this.army.level,
                    link: await TextEditor.enrichHTML(createUUIDLink(t.uuid)),
                    level,
                    dc: getLevelBasedDC(level),
                    name: t.name,
                    uuid: t.uuid,
                };
            }));
    }
}

export async function showArmyTacticsBrowser(options: ArmyTacticsBrowserOptions): Promise<void> {
    new ArmyTacticsBrowserApp(options).render(true);
}

File: src/jsMain/kotlin/at/posselt/pfrpg2e/kingdom/dialogs/ArmyTacticsBrowser.kt
package at.posselt.pfrpg2e.kingdom.dialogs

import at.posselt.pfrpg2e.app.HandlebarsRenderContext
import at.posselt.pfrpg2e.app.forms.SimpleApp
import at.posselt.pfrpg2e.data.checks.getLevelBasedDC
import at.posselt.pfrpg2e.kingdom.KingdomActor
import at.posselt.pfrpg2e.kingdom.KingdomData
import at.posselt.pfrpg2e.kingdom.armies.getAllAvailableArmyTactics
import at.posselt.pfrpg2e.kingdom.armies.getSelectedArmies
import at.posselt.pfrpg2e.kingdom.armies.hasTactic
import at.posselt.pfrpg2e.kingdom.armies.isArmyTactic
import at.posselt.pfrpg2e.kingdom.getActivity
import at.posselt.pfrpg2e.takeIfInstance
import at.posselt.pfrpg2e.utils.awaitAll
import at.posselt.pfrpg2e.utils.buildPromise
import at.posselt.pfrpg2e.utils.buildUuid
import at.posselt.pfrpg2e.utils.launch
import com.foundryvtt.core.Game
import com.foundryvtt.core.applications.api.HandlebarsRenderOptions
import com.foundryvtt.core.ui
import com.foundryvtt.core.ui.TextEditor
import com.foundryvtt.pf2e.actor.PF2EArmy
import com.foundryvtt.pf2e.item.PF2ECampaignFeature
import com.foundryvtt.pf2e.item.itemFromUuid
import js.objects.JsPlainObject
import kotlinx.coroutines.await
import org.w3c.dom.HTMLElement
import org.w3c.dom.get
import org.w3c.dom.pointerevents.PointerEvent
import kotlin.js.Promise

@JsPlainObject
private external interface ArmyTacticContext {
    val link: String
    val level: Int
    val dc: Int
    val enabled: Boolean
    val uuid: String
}


@JsPlainObject
private external interface ArmyTacticsContext : HandlebarsRenderContext {
    val tactics: Array<ArmyTacticContext>
}

private class ArmyTacticsBrowser(
    private val game: Game,
    private val kingdomActor: KingdomActor,
    private val kingdom: KingdomData,
    private val army: PF2EArmy,
) : SimpleApp<ArmyTacticsContext>(
    title = "Learnable Tactics: ${army.name}",
    template = "applications/kingdom/army-tactics-browser.hbs",
    id = "kmArmyTactics-${kingdomActor.uuid}",
    width = 600,
) {
    override fun _onClickAction(event: PointerEvent, target: HTMLElement) {
        when (target.dataset["action"]) {
            "train-tactic" -> {
                buildPromise {
                    val uuid = target.dataset["uuid"] as String
                    trainTactic(uuid)
                }
            }
        }
    }

    override fun _preparePartContext(
        partId: String,
        context: HandlebarsRenderContext,
        options: HandlebarsRenderOptions
    ): Promise<ArmyTacticsContext> = buildPromise {
        val parent = super._preparePartContext(partId, context, options).await()
        val tactics = game.getAllAvailableArmyTactics()
            .asSequence()
            .filter { it.system.traits.value.contains(army.system.traits.type) && !army.hasTactic(it) }
            .sortedWith(compareBy<PF2ECampaignFeature> { it.level }.thenBy { it.name })
            .map {
                buildPromise {
                    val level = it.level
                    val dc = getLevelBasedDC(level)
                    val link = TextEditor.enrichHTML(buildUuid(it.uuid, it.name)).await()
                    ArmyTacticContext(
                        link = link,
                        level = level,
                        dc = dc,
                        enabled = level <= army.level,
                        uuid = it.uuid,
                    )
                }
            }
            .toList()
            .awaitAll()
            .toTypedArray()
        ArmyTacticsContext(
            partId = parent.partId,
            tactics = tactics
        )
    }

    private suspend fun trainTactic(uuid: String) {
        val item = itemFromUuid(uuid)
            ?.takeIfInstance<PF2ECampaignFeature>()
            ?.takeIf { it.isArmyTactic }
        if (item != null) {
            val activity = this.kingdom.getActivity("train-army")
            checkNotNull(activity) {
                "Could not find train-army activity"
            }
            val degreeMessage = "<b>Learned Tactic</b>: ${buildUuid(item.uuid, item.name)}"
            kingdomCheckDialog(
                game = this.game,
                kingdom = this.kingdom,
                kingdomActor = this.kingdomActor,
                check = CheckType.PerformActivity(activity),
                afterRoll = {close()},
                degreeMessages = DegreeMessages(
                    criticalSuccess = degreeMessage,
                    success = degreeMessage,
                ),
                overrideDc = getLevelBasedDC(item.level),
            )
        }
    }
}

fun armyTacticsBrowser(game: Game, kingdomActor: KingdomActor, kingdom: KingdomData) {
    val selectedArmies = game.getSelectedArmies()
    val army = selectedArmies.firstOrNull()
    if (army == null || selectedArmies.size > 1) {
        ui.notifications.error("Please select a single army on the scene")
    } else {
        ArmyTacticsBrowser(game, kingdomActor, kingdom, army).launch()
    }
}